jwql/instrument_monitors/common_monitors/dark_monitor.py

```diff
@@ -0,0 +1,988 @@
+#! /usr/bin/env python
+
+"""This module contains code for the dark current monitor, which
+performs some basic analysis to check whether the dark current behavior
+for the most recent set of input files is consistent with that from
+past files.
+
+If enough new files for a given instrument/aperture combination
+(currently the files must be identified as dark current files in the
+``exp_type`` header keyword) are present in the filesystem at the time
+the ``dark_monitor`` is called, the files are first run through the the
+appropriate pipeline steps to produce slope images.
+
+A mean slope image as well as a standard deviation slope image is
+created by sigma-clipping on a pixel by pixel basis. The mean and
+standard deviation images are saved to a fits file, the name of which
+is entered into the ``<Instrument>DarkCurrent`` database table.
+
+The mean slope image is then normalized by an existing baseline slope
+image. New hot pixels are identified as those with normalized signal
+rates above a ``hot_threshold`` value. Similarly, pixels with
+normalized signal rates below a ``dead_threshold`` are flagged as new
+dead pixels.
+
+The standard deviation slope image is normalized by a baseline
+(historical) standard deviation image. Pixels with normalized values
+above a noise threshold are flagged as newly noisy pixels.
+
+New hot, dead, and noisy pixels are saved to the ``DarkPixelStats``
+database table.
+
+Next, the dark current in the mean slope image is examined. A histogram
+of the slope values is created for the pixels in each amplifier, as
+well as for all pixels on the detector. In all cases, a Gaussian is fit
+to the histogram. Currently for NIRCam and NIRISS, a double Gaussian is
+also fit to the histogram from the entire detector.
+
+The histogram itself as well as the best-fit Gaussian and double
+Gaussian parameters are saved to the DarkDarkCurrent database table.
+
+
+Author
+------
+
+    - Bryan Hilbert
+
+Use
+---
+
+    This module can be used from the command line as such:
+
+    ::
+
+        python dark_monitor.py
+"""
+
+from copy import copy, deepcopy
+import datetime
+import logging
+import os
+
+from astropy.io import ascii, fits
+from astropy.modeling import models
+from astropy.time import Time
+import numpy as np
+from pysiaf import Siaf
+from sqlalchemy import func, Table
+from sqlalchemy.sql.expression import and_
+
+from jwql.database.database_interface import base, session
+from jwql.database.database_interface import NIRCamDarkQueries, NIRCamDarkPixelStats, NIRCamDarkDarkCurrent
+from jwql.database.database_interface import NIRISSDarkQueries, NIRISSDarkPixelStats, NIRISSDarkDarkCurrent
+from jwql.database.database_interface import MIRIDarkQueries, MIRIDarkPixelStats, MIRIDarkDarkCurrent
+from jwql.database.database_interface import NIRSpecDarkQueries, NIRSpecDarkPixelStats, NIRSpecDarkDarkCurrent
+from jwql.database.database_interface import FGSDarkQueries, FGSDarkPixelStats, FGSDarkDarkCurrent
```

Comment:
I think that's ok.  See the `identify_tables` function to see where the tables are used. PyCharm probably isn't picking up on them being used since they are wrapped in the `literal_eval` function. It could be switched to call `monitor_orm_factory` on the three text files for the given instrument. I guess the downside of that is that you would then be recreating the ORM many times, since the dark monitor loops over instrument/aperture combinations.

Reply:
Ah, I hadn't thought about the fact that the ORM would be re-made every time. That's a good enough reason for me to keep it the way it is now.

(I haven't seen that stuff with `eval` before, very neat!)